File: frontend/apps/ios/Pantopus/Features/Homes/Guests/AddGuestFormViewModel.swift
//
//  AddGuestFormViewModel.swift
//  Pantopus
//
//  A13.1 — Backs the Add Guest form (issue a short-term guest pass for a
//  home). Built on the shared `FormShell` archetype: name + contact +
//  welcome are tracked as `FormFieldState`s; duration (single-select) and
//  allowed-areas (multi-select) are enum-ish chip selections held
//  directly.
//
//  `submit()` issues the pass via `POST /api/homes/:id/guest-passes`
//  (route `backend/routes/homeIam.js:667`), raises a success toast
//  ("Pass sent to <name>") and publishes `createdShare` — the one-time
//  share token composed into a viewer link. The host offers the OS share
//  sheet (RN parity: `src/app/homes/[id]/share.tsx:60-82`) and then calls
//  `acknowledgeShare()`, which flips `shouldDismiss` so the modal pops.
//  The contact, welcome note, and allowed-area chips are UI affordances
//  the create endpoint doesn't model, so they stay local.
//

import Foundation
import Observation
import SwiftUI

@Observable
@MainActor
public final class AddGuestFormViewModel {
    // MARK: - Bound state

    public private(set) var nameField: FormFieldState
    public private(set) var contactField: FormFieldState
    public private(set) var welcomeField: FormFieldState

    /// Selected duration chip id (single-select). `nil` until the user
    /// picks one. `"custom"` opens the date-range sheet.
    public var duration: String?

    /// Selected allowed-area chip ids (multi-select, optional).
    public var selectedAreas: Set<String> = []

    /// Custom date range, populated when the user commits the picker.
    public private(set) var customStart: Date?
    public private(set) var customEnd: Date?

    public private(set) var isSaving = false
    public var toast: ToastMessage?
    public private(set) var shouldDismiss = false

    /// Set once the pass is created. Carries the raw one-time `token`
    /// (composed into a public viewer URL) so the host can fire the OS
    /// share sheet. Cleared by `acknowledgeShare()`.
    public private(set) var createdShare: GuestPassShare?

    // MARK: - Inputs

    public let homeId: String
    public let homeContext: AddGuestSampleData.HomeContext
    public let durationOptions = AddGuestSampleData.durationOptions
    public let areaOptions = AddGuestSampleData.areaOptions
    public let welcomeMaxLength = AddGuestSampleData.welcomeMaxLength

    private let onSent: (String) -> Void
    private let api: APIClient

    init(
        homeId: String,
        api: APIClient = .shared,
        onSent: @escaping (String) -> Void = { _ in }
    ) {
        self.homeId = homeId
        self.api = api
        homeContext = AddGuestSampleData.homeContext(for: homeId)
        nameField = FormFieldState(id: "name", originalValue: "")
        contactField = FormFieldState(id: "contact", originalValue: "")
        welcomeField = FormFieldState(id: "welcome", originalValue: "")
        self.onSent = onSent
    }

    // MARK: - Aggregate

    /// Required: name non-empty, contact valid (email OR phone), duration
    /// chosen. Areas + welcome are optional.
    public var isValid: Bool {
        !trimmedName.isEmpty
            && Self.isContactValid(contactField.value)
            && duration != nil
    }

    /// Any input touched — drives the dirty-close confirm in `FormShell`.
    public var isDirty: Bool {
        !trimmedName.isEmpty
            || !contactField.value.isEmpty
            || duration != nil
            || !selectedAreas.isEmpty
            || !welcomeField.value.isEmpty
    }

    // MARK: - Field states (drive the text-field visuals)

    public var nameFieldState: PantopusFieldState {
        (nameField.touched && !trimmedName.isEmpty) ? .valid : .default
    }

    public var contactFieldState: PantopusFieldState {
        if let error = contactField.error { return .error(error) }
        if contactField.touched, Self.isContactValid(contactField.value) { return .valid }
        return .default
    }

    // MARK: - Hints

    /// Italic helper under the duration chips. Mirrors the design copy.
    public var durationHint: String {
        switch duration {
        case "2h": "Two-hour pass · auto-revokes after"
        case "today": "Today until 11:59 PM · auto-revokes after"
        case "weekend": "Sat 12:00 AM → Sun 11:59 PM · auto-revokes after"
        case AddGuestSampleData.durationCustomId:
            if let start = customStart, let end = customEnd {
                "\(Self.format(start)) → \(Self.format(end)) · auto-revokes after"
            } else {
                "Pick a custom date range"
            }
        default: "Pick how long the pass is good for."
        }
    }

    /// Italic helper under the allowed-areas chips.
    public var areasHint: String {
        guard !selectedAreas.isEmpty else { return "Front door only, unless you add more." }
        let possessive = firstName.map { "\($0)'s" } ?? "Their"
        return "\(possessive) pass unlocks only what you pick."
    }

    /// First word of the entered name, if any — used for the toast and
    /// the areas-hint possessive.
    public var firstName: String? {
        trimmedName.split(separator: " ").first.map(String.init)
    }

    private var trimmedName: String {
        nameField.value.trimmingCharacters(in: .whitespacesAndNewlines)
    }

    // MARK: - Field updates

    public func updateName(_ value: String) {
        nameField.value = value
        nameField.touched = true
    }

    public func updateContact(_ value: String) {
        contactField.value = value
        contactField.touched = true
        contactField.error = Self.validateContact(value)
    }

    public func updateWelcome(_ value: String) {
        welcomeField.value = String(value.prefix(welcomeMaxLength))
        welcomeField.touched = true
    }

    // MARK: - Custom duration range

    public func setCustomRange(_ start: Date, _ end: Date) {
        customStart = start
        customEnd = end
        duration = AddGuestSampleData.durationCustomId
    }

    public func clearCustomRange() {
        customStart = nil
        customEnd = nil
        if duration == AddGuestSampleData.durationCustomId { duration = nil }
    }

    // MARK: - Submit

    public func submit() async {
        guard isValid, !isSaving else { return }
        isSaving = true
        // `label` carries the guest's name; the time window comes from the
        // selected duration chip. Contact, welcome note, and allowed-area
        // chips are UI affordances the create endpoint doesn't model, so
        // they stay local (a backend follow-up would persist them).
        let window = guestPassWindow()
        let request = CreateGuestPassRequest(
            label: trimmedName,
            kind: "guest",
            durationHours: window.durationHours,
            startAt: window.startAt,
            endAt: window.endAt
        )
        let response: CreateGuestPassResponse
        do {
            response = try await api.request(
                HomesEndpoints.createGuestPass(homeId: homeId, request: request),
                as: CreateGuestPassResponse.self
            )
        } catch {
            isSaving = false
            toast = ToastMessage(
                text: (error as? APIError)?.errorDescription
                    ?? "Couldn't issue the pass. Try again.",
                kind: .error
            )
            return
        }
        isSaving = false
        let name = firstName ?? "your guest"
        toast = ToastMessage(text: "Pass sent to \(name)", kind: .success)
        onSent(name)
        // The raw token is returned exactly once (homeIam.js:762-767) —
        // this is the only moment a shareable link can be built.
        createdShare = GuestPassShare(
            id: response.pass.id,
            guestName: firstName ?? "",
            urlString: GuestPassShare.url(forToken: response.token)
        )
    }

    /// Called by the host once the user has either shared the new pass or
    /// declined. Clears the offer and asks the host to pop the form.
    public func acknowledgeShare() {
        createdShare = nil
        shouldDismiss = true
    }

    // MARK: - Guest-pass window

    /// The backend resolves `end_at` > `duration_hours` > template
    /// default, so the 2-hour preset sends a relative `durationHours`
    /// while the dated presets send absolute ISO `start`/`end` stamps.
    struct GuestPassWindow {
        var durationHours: Int?
        var startAt: String?
        var endAt: String?
    }

    func guestPassWindow() -> GuestPassWindow {
        switch duration {
        case "2h":
            return GuestPassWindow(durationHours: 2)
        case "today":
            return GuestPassWindow(endAt: Self.isoStartOfTomorrow())
        case "weekend":
            let (start, end) = Self.weekendWindow()
            return GuestPassWindow(startAt: start, endAt: end)
        case AddGuestSampleData.durationCustomId:
            // Day-precision boundaries (parity with Android's day picker):
            // start of the first day → start of the day after the last.
            return GuestPassWindow(
                startAt: customStart.map(Self.isoStartOfDay),
                endAt: customEnd.map(Self.isoStartOfNextDay)
            )
        default:
            // `isValid` guarantees a duration is chosen before submit;
            // fall back to the 2-hour preset defensively.
            return GuestPassWindow(durationHours: 2)
        }
    }

    private static func iso(_ date: Date) -> String {
        let formatter = ISO8601DateFormatter()
        formatter.formatOptions = [.withInternetDateTime]
        return formatter.string(from: date)
    }

    private static func isoStartOfDay(_ date: Date) -> String {
        iso(Calendar.current.startOfDay(for: date))
    }

    private static func isoStartOfNextDay(_ date: Date) -> String {
        let calendar = Calendar.current
        let startOfDay = calendar.startOfDay(for: date)
        let next = calendar.date(byAdding: .day, value: 1, to: startOfDay) ?? startOfDay
        return iso(next)
    }

    /// Local next-midnight — the exclusive end of "today".
    private static func isoStartOfTomorrow() -> String {
        let calendar = Calendar.current
        let startOfToday = calendar.startOfDay(for: Date())
        let startOfTomorrow = calendar.date(byAdding: .day, value: 1, to: startOfToday) ?? startOfToday
        return iso(startOfTomorrow)
    }

    /// Upcoming (or current) Sat 00:00 → Mon 00:00 window.
    private static func weekendWindow() -> (String, String) {
        let calendar = Calendar.current
        let startOfToday = calendar.startOfDay(for: Date())
        // `.weekday`: 1 = Sunday … 7 = Saturday.
        let weekday = calendar.component(.weekday, from: startOfToday)
        let daysUntilSaturday = (7 - weekday) % 7
        let saturday = calendar.date(byAdding: .day, value: daysUntilSaturday, to: startOfToday) ?? startOfToday
        let monday = calendar.date(byAdding: .day, value: 2, to: saturday) ?? saturday
        return (iso(saturday), iso(monday))
    }

    public func acknowledgeDismiss() {
        shouldDismiss = false
    }

    // MARK: - Validation (static — testable without the view layer)

    /// Returns an error string when a non-empty contact is neither a
    /// valid email nor a valid phone; `nil` when empty (required-ness is
    /// enforced by `isValid`) or valid.
    static func validateContact(_ raw: String) -> String? {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty { return nil }
        return isContactValid(trimmed) ? nil : "Enter a valid email or phone number."
    }

    static func isContactValid(_ raw: String) -> Bool {
        let trimmed = raw.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !trimmed.isEmpty else { return false }
        return matchesEmail(trimmed) || matchesPhone(trimmed)
    }

    static func matchesEmail(_ value: String) -> Bool {
        let pattern = "^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$"
        return value.range(of: pattern, options: .regularExpression) != nil
    }

    static func matchesPhone(_ value: String) -> Bool {
        let allowed = CharacterSet(charactersIn: "0123456789 +()-.")
        guard value.unicodeScalars.allSatisfy({ allowed.contains($0) }) else { return false }
        let digits = value.filter(\.isNumber)
        return digits.count >= 7 && digits.count <= 15
    }

    private static func format(_ date: Date) -> String {
        let formatter = DateFormatter()
        formatter.locale = Locale(identifier: "en_US_POSIX")
        formatter.dateFormat = "MMM d"
        return formatter.string(from: date)
    }
}

File: frontend/apps/android/app/src/main/java/app/pantopus/android/ui/screens/homes/guests/AddGuestFormViewModel.kt
@file:Suppress("PackageNaming", "MagicNumber", "TooManyFunctions")

package app.pantopus.android.ui.screens.homes.guests

import androidx.lifecycle.SavedStateHandle
import androidx.lifecycle.ViewModel
import androidx.lifecycle.viewModelScope
import app.pantopus.android.data.api.models.homes.CreateGuestPassRequest
import app.pantopus.android.data.api.net.NetworkResult
import app.pantopus.android.data.homes.HomeGuestPassesRepository
import app.pantopus.android.ui.components.ChipPickerOption
import app.pantopus.android.ui.screens.shared.form.FormFieldState
import dagger.hilt.android.lifecycle.HiltViewModel
import kotlinx.coroutines.flow.MutableStateFlow
import kotlinx.coroutines.flow.StateFlow
import kotlinx.coroutines.flow.asStateFlow
import kotlinx.coroutines.flow.update
import kotlinx.coroutines.launch
import java.time.DayOfWeek
import java.time.LocalDate
import java.time.ZoneId
import java.time.temporal.TemporalAdjusters
import javax.inject.Inject

/** Nav-arg key for the home id consumed via [SavedStateHandle]. */
const val ADD_GUEST_HOME_ID_KEY = "homeId"

/** Tone + text payload the form turns into a transient toast. */
data class GuestToast(
    val text: String,
    val isError: Boolean,
)

/**
 * Aggregate UI state for the Add Guest form. Mirrors iOS
 * `AddGuestFormViewModel`'s exposed state: name / contact / welcome are
 * tracked as [FormFieldState]s, duration is a single-select chip id, and
 * allowed-areas is a set of chip ids.
 */
data class AddGuestUiState(
    val homeTitle: String = "",
    val homeSubtitle: String = "",
    val nameField: FormFieldState = FormFieldState(id = "name"),
    val contactField: FormFieldState = FormFieldState(id = "contact"),
    val welcomeField: FormFieldState = FormFieldState(id = "welcome"),
    val duration: String? = null,
    val selectedAreas: Set<String> = emptySet(),
    val customStartLabel: String? = null,
    val customEndLabel: String? = null,
    val customStartEpochDay: Long? = null,
    val customEndEpochDay: Long? = null,
    val isSaving: Boolean = false,
    val toast: GuestToast? = null,
    val shouldDismiss: Boolean = false,
    /**
     * A13.6 — set once the pass is created. Carries the raw one-time
     * `token` (composed into a public viewer URL) so the screen can fire
     * the OS share sheet. Cleared by `acknowledgeShare()`.
     */
    val createdShare: GuestPassShare? = null,
) {
    val durationOptions: List<ChipPickerOption> get() = AddGuestSampleData.durationOptions
    val areaOptions: List<ChipPickerOption> get() = AddGuestSampleData.areaOptions
    val welcomeMaxLength: Int get() = AddGuestSampleData.WELCOME_MAX_LENGTH

    /** First word of the entered name, if any. */
    val firstName: String?
        get() = nameField.value.trim().split(" ").firstOrNull()?.takeIf { it.isNotEmpty() }

    /** Required: name non-empty, contact valid (email OR phone), duration chosen. */
    val isValid: Boolean
        get() =
            nameField.value.trim().isNotEmpty() &&
                isGuestContactValid(contactField.value) &&
                duration != null

    /** Any input touched — drives the dirty-close confirm in `FormShell`. */
    val isDirty: Boolean
        get() =
            nameField.value.trim().isNotEmpty() ||
                contactField.value.isNotEmpty() ||
                duration != null ||
                selectedAreas.isNotEmpty() ||
                welcomeField.value.isNotEmpty()

    /** Italic helper under the duration chips. Mirrors the design copy. */
    val durationHint: String
        get() =
            when (duration) {
                "2h" -> "Two-hour pass · auto-revokes after"
                "today" -> "Today until 11:59 PM · auto-revokes after"
                "weekend" -> "Sat 12:00 AM → Sun 11:59 PM · auto-revokes after"
                AddGuestSampleData.DURATION_CUSTOM_ID ->
                    if (customStartLabel != null && customEndLabel != null) {
                        "$customStartLabel → $customEndLabel · auto-revokes after"
                    } else {
                        "Pick a custom date range"
                    }
                else -> "Pick how long the pass is good for."
            }

    /** Italic helper under the allowed-areas chips. */
    val areasHint: String
        get() =
            if (selectedAreas.isEmpty()) {
                "Front door only, unless you add more."
            } else {
                val possessive = firstName?.let { "$it's" } ?: "Their"
                "$possessive pass unlocks only what you pick."
            }
}

/**
 * A13.1 — Add Guest form view-model. [submit] issues the pass via
 * `POST /api/homes/:id/guest-passes` (route `backend/routes/homeIam.js:667`),
 * raises a success toast ("Pass sent to <name>") and publishes
 * [AddGuestUiState.createdShare] — the one-time share token composed into a
 * viewer link. The screen offers the OS share sheet (RN parity:
 * `src/app/homes/[id]/share.tsx:60-82`) and then calls [acknowledgeShare],
 * which flips [AddGuestUiState.shouldDismiss] so the host pops the form.
 * The contact, welcome note, and allowed-area chips are UI affordances the
 * create endpoint doesn't model, so they stay local.
 */
@HiltViewModel
class AddGuestFormViewModel
    @Inject
    constructor(
        savedStateHandle: SavedStateHandle,
        private val guestPassesRepo: HomeGuestPassesRepository,
    ) : ViewModel() {
        private val homeId: String = savedStateHandle.get<String>(ADD_GUEST_HOME_ID_KEY) ?: ""

        private val _state =
            MutableStateFlow(
                AddGuestSampleData.homeContext(homeId).let { ctx ->
                    AddGuestUiState(homeTitle = ctx.title, homeSubtitle = ctx.subtitle)
                },
            )
        val state: StateFlow<AddGuestUiState> = _state.asStateFlow()

        fun updateName(value: String) {
            _state.update { it.copy(nameField = it.nameField.copy(value = value, touched = true)) }
        }

        fun updateContact(value: String) {
            _state.update {
                it.copy(
                    contactField =
                        it.contactField.copy(
                            value = value,
                            touched = true,
                            error = validateGuestContact(value),
                        ),
                )
            }
        }

        fun updateWelcome(value: String) {
            val clipped = value.take(AddGuestSampleData.WELCOME_MAX_LENGTH)
            _state.update { it.copy(welcomeField = it.welcomeField.copy(value = clipped, touched = true)) }
        }

        fun setDuration(id: String?) {
            _state.update { it.copy(duration = id) }
        }

        fun setAreas(areas: Set<String>) {
            _state.update { it.copy(selectedAreas = areas) }
        }

        fun setCustomRange(
            startLabel: String,
            endLabel: String,
            startEpochDay: Long,
            endEpochDay: Long,
        ) {
            _state.update {
                it.copy(
                    duration = AddGuestSampleData.DURATION_CUSTOM_ID,
                    customStartLabel = startLabel,
                    customEndLabel = endLabel,
                    customStartEpochDay = startEpochDay,
                    customEndEpochDay = endEpochDay,
                )
            }
        }

        fun clearCustomRange() {
            _state.update {
                val nextDuration =
                    if (it.duration == AddGuestSampleData.DURATION_CUSTOM_ID) null else it.duration
                it.copy(
                    duration = nextDuration,
                    customStartLabel = null,
                    customEndLabel = null,
                    customStartEpochDay = null,
                    customEndEpochDay = null,
                )
            }
        }

        fun submit() {
            val current = _state.value
            if (!current.isValid || current.isSaving) return
            _state.update { it.copy(isSaving = true) }
            viewModelScope.launch {
                val window = guestPassWindow(current)
                val request =
                    CreateGuestPassRequest(
                        label = current.nameField.value.trim(),
                        kind = "guest",
                        durationHours = window.durationHours,
                        startAt = window.startAt,
                        endAt = window.endAt,
                    )
                when (val result = guestPassesRepo.create(homeId, request)) {
                    is NetworkResult.Success -> {
                        val firstName = _state.value.firstName
                        val name = firstName ?: "your guest"
                        // The raw token is returned exactly once
                        // (homeIam.js:762-767) — this is the only moment a
                        // shareable link can be built.
                        val share =
                            GuestPassShare(
                                id = result.data.pass.id,
                                guestName = firstName.orEmpty(),
                                url = GuestPassShare.urlForToken(result.data.token),
                            )
                        _state.update {
                            it.copy(
                                isSaving = false,
                                toast = GuestToast("Pass sent to $name", isError = false),
                                createdShare = share,
                            )
                        }
                    }
                    is NetworkResult.Failure -> {
                        _state.update {
                            it.copy(
                                isSaving = false,
                                toast =
                                    GuestToast(
                                        result.error.message.ifEmpty { "Couldn't issue the pass. Try again." },
                                        isError = true,
                                    ),
                            )
                        }
                    }
                }
            }
        }

        fun dismissToast() {
            _state.update { it.copy(toast = null) }
        }

        fun acknowledgeDismiss() {
            _state.update { it.copy(shouldDismiss = false) }
        }

        /**
         * A13.6 — called by the screen once the user has either shared the
         * new pass or declined. Clears the offer and asks the host to pop
         * the form.
         */
        fun acknowledgeShare() {
            _state.update { it.copy(createdShare = null, shouldDismiss = true) }
        }

        // ─── Guest-pass window ──────────────────────────────────────

        private data class GuestPassWindow(
            val durationHours: Int? = null,
            val startAt: String? = null,
            val endAt: String? = null,
        )

        /**
         * Maps the selected duration chip to the backend timing fields.
         * The handler resolves `end_at` > `duration_hours` > template
         * default, so the 2-hour preset sends a relative duration while
         * the dated presets send absolute ISO `start`/`end` stamps.
         * Mirrors iOS `AddGuestFormViewModel.guestPassWindow()`.
         */
        private fun guestPassWindow(state: AddGuestUiState): GuestPassWindow {
            val zone = ZoneId.systemDefault()
            return when (state.duration) {
                "2h" -> GuestPassWindow(durationHours = 2)
                "today" ->
                    GuestPassWindow(
                        endAt = LocalDate.now(zone).plusDays(1).atStartOfDay(zone).toInstant().toString(),
                    )
                "weekend" -> {
                    val saturday = LocalDate.now(zone).with(TemporalAdjusters.nextOrSame(DayOfWeek.SATURDAY))
                    val monday = saturday.plusDays(2)
                    GuestPassWindow(
                        startAt = saturday.atStartOfDay(zone).toInstant().toString(),
                        endAt = monday.atStartOfDay(zone).toInstant().toString(),
                    )
                }
                AddGuestSampleData.DURATION_CUSTOM_ID -> {
                    val start = state.customStartEpochDay
                    val end = state.customEndEpochDay
                    if (start != null && end != null) {
                        GuestPassWindow(
                            startAt = LocalDate.ofEpochDay(start).atStartOfDay(zone).toInstant().toString(),
                            endAt = LocalDate.ofEpochDay(end).plusDays(1).atStartOfDay(zone).toInstant().toString(),
                        )
                    } else {
                        GuestPassWindow(durationHours = 2)
                    }
                }
                else -> GuestPassWindow(durationHours = 2)
            }
        }
    }

// ─── Validation — top-level so the UiState getters + VM share them ──────

internal fun validateGuestContact(raw: String): String? {
    val trimmed = raw.trim()
    if (trimmed.isEmpty()) return null
    return if (isGuestContactValid(trimmed)) null else "Enter a valid email or phone number."
}

internal fun isGuestContactValid(raw: String): Boolean {
    val trimmed = raw.trim()
    if (trimmed.isEmpty()) return false
    return matchesGuestEmail(trimmed) || matchesGuestPhone(trimmed)
}

private val guestEmailRegex = Regex("^[A-Za-z0-9._%+-]+@[A-Za-z0-9.-]+\\.[A-Za-z]{2,}$")
private val guestPhoneAllowed = Regex("^[0-9 +()\\-.]+$")

private fun matchesGuestEmail(value: String): Boolean = guestEmailRegex.matches(value)

private fun matchesGuestPhone(value: String): Boolean {
    if (!guestPhoneAllowed.matches(value)) return false
    val digits = value.count { it.isDigit() }
    return digits in 7..15
}
